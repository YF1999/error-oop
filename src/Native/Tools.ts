export function appendInnerErrorStack(currStack?: string, innerError?: Error): string | undefined {
    return innerError && innerError.stack
        ? `${currStack ?? 'No stack.'}\n--- inner error ---\n${innerError.stack}`
        : currStack;
}

export function setNonEnumerable(obj: object, property: string): void {
    Reflect.defineProperty(obj, property, {
        configurable: true,
        enumerable: false,
        writable: true,
    });
}
