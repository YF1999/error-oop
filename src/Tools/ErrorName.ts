export enum ErrorName {
    // native errors

    NativeError = 'NativeError',
    NativeRangeError = 'NativeRangeError',
    NativeReferenceError = 'NativeReferenceError',
    NativeSyntaxError = 'NativeSyntaxError',
    NativeTypeError = 'NativeTypeError',
    NativeURIError = 'NativeURIError',

    // more oop errors

    AlreadyInUseError = 'AlreadyInUseError',
    ArgumentError = 'ArgumentError',
    ArgumentNullError = 'ArgumentNullError',
    ArgumentOutOfRangeError = 'ArgumentOutOfRangeError',
    AuthenticationError = 'AuthenticationError',
    AuthenticationRequiredError = 'AuthenticationRequiredError',
    ConnectionError = 'ConnectionError',
    DataError = 'DataError',
    DirectoryNotFoundError = 'DirectoryNotFoundError',
    DriveNotFoundError = 'DriveNotFoundError',
    EndOfStreamError = 'EndOfStreamError',
    FileLoadError = 'FileLoadError',
    FileNotFoundError = 'FileNotFoundError',
    IndexOutOfRangeError = 'IndexOutOfRangeError',
    InvalidDataError = 'InvalidDataError',
    InvalidOperationError = 'InvalidOperationError',
    IOError = 'IOError',
    NotFoundError = 'NotFoundError',
    NotImplementedError = 'NotImplementedError',
    NotPermittedError = 'NotPermittedError',
    NotSupportedError = 'NotSupportedError',
    OperationCanceledError = 'OperationCanceledError',
    OutOfMemoryError = 'OutOfMemoryError',
    PathTooLongError = 'PathTooLongError',
    SocketError = 'SocketError',
    StackOverflowError = 'StackOverflowError',
    TimeoutError = 'TimeoutError',
    TransactionError = 'TransactionError',
    UnauthorizedAccessError = 'UnauthorizedAccessError',
    UnauthorizedError = 'UnauthorizedError',
}
