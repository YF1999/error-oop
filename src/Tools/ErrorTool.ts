import { IArgumentErrorOptions } from '../ArgumentError';
import { IIOErrorOptions } from '../IO/IOError';
import { IErrorOptions, IStandardArguments } from '../Native';

export const ErrorTool = {
    parseStandardArguments<T extends IErrorOptions>(...args: IStandardArguments<T>): IErrorOptions {
        let options: IErrorOptions = {};

        if (args.length === 1) {
            if (typeof args[0] === 'object') {
                [options] = args;
            } else {
                [options.message] = args;
            }
        } else if (args.length === 2) {
            [options.message, options.innerError] = args;
        }

        return options;
    },

    parseArgumentArguments<T extends IArgumentErrorOptions>(
        ...args: [] | [T] | [string, Error?] | [string, string, Error?] | [string, string, unknown, Error?]
    ): IErrorOptions & {
        paramName?: string;
        actualValue?: unknown;
        noMessageSuffix?: boolean;
    } {
        let options: IErrorOptions & { paramName?: string; actualValue?: unknown; noMessageSuffix?: boolean } = {};

        if (args.length === 1) {
            if (typeof args[0] === 'object') {
                options = args[0];
            } else {
                options.message = args[0];
            }
        } else if (args.length === 2) {
            if (typeof args[1] === 'object') {
                [options.message, options.innerError] = args;
            } else {
                [options.message, options.paramName] = args;
            }
        } else if (args.length === 3) {
            if (args[2] instanceof Error) {
                [options.message, options.paramName, options.innerError] = args;
            } else {
                [options.message, options.paramName, options.actualValue] = args;
            }
        } else if (args.length === 4) {
            [options.message, options.paramName, options.actualValue, options.innerError] = args;
        }

        return options;
    },

    parseEntityArguments<T extends IIOErrorOptions>(
        ...args: [] | [T] | [string, Error?] | [string, string, Error?]
    ): IIOErrorOptions & { entityName?: string } {
        let options: IIOErrorOptions & { entityName?: string } = {};

        if (args.length === 1) {
            if (typeof args[0] === 'object') {
                options = args[0];
            } else {
                options.message = args[0];
            }
        } else if (args.length === 2) {
            if (typeof args[1] === 'object') {
                [options.message, options.innerError] = args;
            } else {
                [options.message, options.entityName] = args;
            }
        } else if (args.length === 3) {
            [options.message, options.entityName, options.innerError] = args;
        }

        return options;
    },

    parseFileArguments<T extends IIOErrorOptions>(
        ...args: [] | [T] | [string, Error?] | [string, string, Error?]
    ): IIOErrorOptions & { fileName?: string } {
        let options: IIOErrorOptions & { fileName?: string } = {};

        if (args.length === 1) {
            if (typeof args[0] === 'object') {
                options = args[0];
            } else {
                options.message = args[0];
            }
        } else if (args.length === 2) {
            if (typeof args[1] === 'object') {
                [options.message, options.innerError] = args;
            } else {
                [options.message, options.fileName] = args;
            }
        } else if (args.length === 3) {
            [options.message, options.fileName, options.innerError] = args;
        }

        return options;
    },
};
