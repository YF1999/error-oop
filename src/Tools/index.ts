export * from './ErrorName';
export * from './ErrorTool';
