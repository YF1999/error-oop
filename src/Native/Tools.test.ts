import { appendInnerErrorStack, setNonEnumerable } from './Tools';

test('Test `appendInnerErrorStack`', () => {
    const error = { stack: 'inner error stack' };

    expect(appendInnerErrorStack(undefined, error as Error)).toMatchSnapshot();
    expect(appendInnerErrorStack('curr error stack', error as Error)).toMatchSnapshot();
});

test('Test `setNonEnumerable`', () => {
    const obj = { value: 'value' };

    expect(Object.keys(obj)).toStrictEqual(['value']);
    setNonEnumerable(obj, 'value');
    expect(Object.keys(obj)).toStrictEqual([]);
});
