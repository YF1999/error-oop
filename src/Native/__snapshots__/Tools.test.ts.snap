// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`appendInnerErrorStack\` 1`] = `
"No stack.
--- inner error ---
inner error stack"
`;

exports[`Test \`appendInnerErrorStack\` 2`] = `
"curr error stack
--- inner error ---
inner error stack"
`;
